File: 2018-Dec-01/BinaryTreeSerialiseAndDeserialise.cpp
#include <cassert>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

struct Node
{
	std::string val;

	Node* left;
	Node* right;

	Node(const std::string& val, Node* left = nullptr, Node* right = nullptr)
		: val(val), left(left), right(right)
	{}

	~Node()
	{
		if (left != nullptr)
		{
			delete left;
		}

		if (right != nullptr)
		{
			delete right;
		}
	}
};

const std::string NULL_NODE = "-";

std::string SerialiseTree(Node* node)
{
	if (node == nullptr)
	{
		return NULL_NODE + " ";
	}

	std::string serialisedTree;

	serialisedTree += node->val + " ";
	serialisedTree += SerialiseTree(node->left);
	serialisedTree += SerialiseTree(node->right);

	return serialisedTree;
}

Node* DeserialiseHelper(std::string& subtree)
{
	std::stringstream scanner(subtree);
	std::string value;

	scanner >> value;

	unsigned int firstIndex = subtree.find_first_of(' ') + 1;
	unsigned int length = subtree.size() - firstIndex;

	subtree = subtree.substr(firstIndex, length);

	if (value == NULL_NODE)
	{
		return nullptr;
	}

	Node* node = new Node(value);
	node->left = DeserialiseHelper(subtree);
	node->right = DeserialiseHelper(subtree);

	return node;
}

Node* DeserialiseTree(const std::string& tree)
{
	std::stringstream scanner(tree);
	std::string treeString(tree);
	std::string value;

	scanner >> value;

	Node* node = new Node(value);
	node = DeserialiseHelper(treeString);
	return node;
}

int main(int argc, char* argv[])
{
	Node* leftLeft = new Node("left.left");
	Node* leftRight = new Node("left.right");
	Node* right = new Node("right");
	Node* left = new Node("left", leftLeft, leftRight);

	Node node("root", left, right);

	std::cout << SerialiseTree(&node) << std::endl;

	assert(DeserialiseTree(SerialiseTree(&node))->left->left->val == "left.left");	// Assertion passes

	std::cin.get();
	
	return 0;
}

File: Days 01 - 10/Day 03/BinaryTreeSerialiseAndDeserialise.cs
using System;
using System.Diagnostics;

namespace DailyCodingProblems
{
	internal class Node
	{
		private const string NullNode = "-";

		public string Value { get; private set; }

		public Node Left { get; private set; }
		public Node Right { get; private set; }

		public Node(string value, Node left = null, Node right = null)
		{
			this.Value = value;
			this.Left = left;
			this.Right = right;
		}

		public string Serialise()
		{
			return SerialiseHelper(this);
		}

		private string SerialiseHelper(Node node)
		{
			if (node == null)
			{
				return NullNode + " ";
			}

			string serialisedTree = node.Value + " ";
			serialisedTree += SerialiseHelper(node.Left);
			serialisedTree += SerialiseHelper(node.Right);

			return serialisedTree;
		}

		public static Node Deserialise(string tree)
		{
			return DeserialiseHelper(tree);
		}

		private static Node DeserialiseHelper(string tree)
		{
			string value = tree.Substring(0, tree.IndexOf(' '));
			string subtree = tree.Substring(tree.IndexOf(' ') + 1);

			if (value == NullNode)
			{
				return null;
			}

			Node node = new Node(value);
			node.Left = DeserialiseHelper(subtree);
			node.Right = DeserialiseHelper(subtree);

			return node;
		}
	}

	internal class Day3
	{
		private static int Main(string[] args)
		{
			Node leftLeft = new Node("left.left");
			Node leftRight = new Node("left.right");
			Node right = new Node("right");
			Node left = new Node("left", leftLeft, leftRight);
			Node node = new Node("root", left, right);

			string decodedNode = Node.Deserialise(node.Serialise()).Left.Left.Value;

			Debug.Assert(decodedNode == "left.left");

			Console.ReadLine();

			return 0;
		}
	}
}